### ISSUE TITLE
Issue where the generated field of my output type is not the same as my POKO entity?

### ISSUE THREAD

**Issue by wojcik-marcin** [leangen/graphql-spqr#462]
### Discussed in https://github.com/leangen/graphql-spqr/discussions/461

<div type='discussions-op-text'>

<sup>Originally posted by **wojcik-marcin** June  5, 2023</sup>
I would like to request some guidance with an issue I am experiencing.

I am making use of the SPQR Spring Boot Starter v0.0.7 within my Kotlin Spring Boot application.

I have a Person data class defined as follows:

``` kotlin
data class Person(
   val name: String,
   val isMarried: Boolean
)
```

I have a @GraphQLMutation endpoint to update the Person

 ``` kotlin
@GraphQLMutation
fun updatePerson(@GraphQLArgument(name = "person") person: Person): Person {
    person.isMarried = true
    return Person
 }
 ```
 
 The GraphQL schema types generated are as follows:
 
 ``` graphql 
type Person {
    name: String!
    married: Boolean!
}

type PersonInput {
    name: String!
    isMarried: Boolean!
}
```

As you can see, the input type generates the property name for `isMarried` correctly, however, I expected the same for the Person type that is being returned as the output type. Instead, it is returning as `married`.

I am using Jackson as my JSON serializer, and adding `@JsonProperty("isMarried")` to the field in the data class doesn't seem to work.

How can I enforce that the output type's `isMarried` property generates correctly to be the same as that of what is defined in the Person data class?

Any assistance in this matter will be greatly appreciated.</div>

**Comment by kaqqao** [comment]
I'm not sure what names Kotlin ends up generating for the getter and setter, but it seems like it does not generate JavaBeans compliant names (which would be 'isIsMarried` or `getIsMarried', and `setIsMarried`). Side note: this is why it's normally preferred in Java to simply name boolean fields `x` and not `isX`, as the related getter will be named `isX()`.
I _guess_ Kotlin generated `isMarried()` and `setIsMarried`, which would explain what SPQR ends up seeing.

Anyway, the simplest would be to append `@GraphQLQuery(name = "isMarried")` on the field, or the getter (maybe via `@get:GraphQLQuery`?). I'm unfortunately not familiar with Kotlin, so can't  say exactly.

The bottom line is that SPQR by default expects JavaBeans spec to be respected, or explicitly overridden via annotations. Alternatively, you can always provide a custom `ResolverBuilder` (the easiest is probably to extend `BeanResolverBuilder`).

**Comment by wojcik-marcin** [comment]
Thanks for your response.

I added the @GraphQLQuery annotation to the isMarried property as follows:

``` kotlin
@GraphQLType(name = "Person")
data class Person(
    val name: String,
    @GraphQLQuery(name = "isMarried")
    var isMarried: Boolean? = null
)
```

That did not seem to work. I am still seeing the generated GraphQL type returning the property as `married`.

I will try with a CustomOperationInfoGenerator and give feedback.

**Comment by wojcik-marcin** [comment]
When I changed the name of the `isMarried` field in to `married`, then the **@GraphQLQuery(name = "isMarried")** annotation successfully renamed the field in the GraphQLType to `isMarried`.

``` kotlin 
@GraphQLType(name = "Person")
data class Person(
    val name: String,
    @GraphQLQuery(name = "isMarried")
    var married: Boolean? = null
)
```

However, if I leave the name of the field in my data class as `isMarried`, then it seems as though the annotation is being ignored

**Comment by IceBlizz6** [comment]
I think i may have figure out something.
The codebase is using jackson to resolve and name fields.
Try to add jackson-module-kotlin dependency to your project.

implementation("com.fasterxml.jackson.module:jackson-module-kotlin:2.9.8")
Let me know if it works for you.
I just tried it and it looks very promising.

**Comment by IceBlizz6** [comment]
Ignore that last post (although you may still want jackson-module-kotlin to solve other issues)
I believe the solution here is to just write a SchemaTransformer.
I made an attempt at this earlier as i had the same problem.

This has not been properly tested, but it may be a starting point for you:
```
class PreserveBooleanFieldName : SchemaTransformer {
	override fun transformField(
		field: GraphQLFieldDefinition,
		operation: Operation,
		operationMapper: OperationMapper,
		buildContext: BuildContext
	): GraphQLFieldDefinition {
		return if (isBoolean(field.type)) {
			GraphQLFieldDefinition
				.newFieldDefinition(field)
				.name(getOriginalName(operation.typedElement))
				.build()
		} else {
			field
		}
	}

	override fun transformInputField(
		field: GraphQLInputObjectField,
		inputField: InputField,
		operationMapper: OperationMapper,
		buildContext: BuildContext
	): GraphQLInputObjectField {
		return if (isBoolean(field.type)) {
			GraphQLInputObjectField
				.newInputObjectField(field)
				.name(getOriginalName(inputField.typedElement))
				.build()
		} else {
			field
		}
	}

	private fun isBoolean(type: GraphQLType): Boolean {
		val typeName = type.toString()
		return typeName == "Boolean" || typeName == "Boolean!"
	}

	private fun getOriginalName(typedElement: TypedElement): String {
		val elements = typedElement.elements
		val method = elements.filterIsInstance<Method>().singleOrNull()
		val field = elements.filterIsInstance<Field>().singleOrNull()
		if (method != null) {
			return method.name
		} else if (field != null) {
			return field.name
		} else {
			error("No elements found")
		}
	}
}
```

@kaqqao Do you have a better suggestion for the isBoolean implementation?
that string equals check doesn't look great.
